strawberryfields/utils.py

```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
+      indices correspond to output-input pairs of the same mode.
+
+
+    Example:
+        This shows the Hong-Ou-Mandel effect by extracting the unitary of a 50/50 beamsplitter, and then
+        computing the output given by one photon at each input (notice the order of the indices: :math:`[out_1, in_1, out_2, in_2,\dots]`).
+        The result tells us that the two photons always emerge together from a random output port and never one per port.
+
+    >>> engine, (A, B) = sf.Engine(num_subsystems=2)
+    >>> with engine:
+    >>>     BSgate(np.pi/4) | (A, B)
+    >>> U = extract_unitary(engine, cutoff_dim=3)
+    >>> print(abs(U[:,1,:,1])**2)
+    [[0.  0.  0.5]
+     [0.  0.  0. ]
+     [0.5 0.  0. ]])
+
+    Args:
+        engine (Engine): the engine containing the circuit
+        cutoff_dim (int): dimension of each index.
+        vectorize_modes (bool): if True, reshape input and output modes in order to return a matrix.
+        backend (str): the backend to build the unitary. ``'fock'`` (default) and ``'tf'`` are supported.
+
+    Returns:
+        array: the numerical array of the unitary circuit
+
+    Raises:
+        TypeError: if the operations used to construct the circuit are not all unitary.
+    """
+
+    if not is_unitary(engine):
+        raise TypeError("The circuit definition contains elements that are not of type Gate")
+
+    if backend not in ('fock', 'tf'):
+        raise ValueError("Only 'fock' and 'tf' backends are supported")
+
+    from copy import deepcopy
+
+    # This is an independent copy of the engine object, with an additional Command at the beginning
+    # of the cmd_queue which creates a ket made of identities.
+    # The core idea is that when we apply the unitary to the "identity" ket
+    # we obtain the unitary itself as the result.
+    _engine = _engine_with_CJ_cmd_queue(deepcopy(engine), cutoff_dim=cutoff_dim)
+
+    _backend = load_backend(backend) # (!) _backend is the object, backend is just 'fock' or 'tf'
+    _backend.begin_circuit(num_subsystems=_engine.num_subsystems, cutoff_dim=cutoff_dim, hbar=_engine.hbar, pure=True)
+    result = _engine.run(_backend, cutoff_dim=cutoff_dim).ket()
+
+    N = _engine.init_num_subsystems
+
+    if vectorize_modes:
+        if backend == 'fock':
+            return np.reshape(result, [cutoff_dim**N, cutoff_dim**N])
+
+        return tf.reshape(result, [cutoff_dim**N, cutoff_dim**N])
+
+    # here we rearrange the indices to go back to the order [in1, out1, in2, out2, etc...]
+    if backend == 'fock':
+        return np.transpose(result, [int(n) for n in np.arange(2*N).reshape((2, N)).T.reshape([-1])])
+
+    return tf.transpose(result, [int(n) for n in np.arange(2*N).reshape((2, N)).T.reshape([-1])])
+
+
+def extract_channel(engine, cutoff_dim: int, representation: str = 'choi', vectorize_modes: bool = False):
+    r"""Returns a numerical array representation of a channel from a queued Engine circuit.
```

Comment:
Because `extract_channel` requires a lot of index manipulations that I would have to rewrite in TensorFlow and duplicate a lot of code. in `extract_unitary` it's easier because it was just a couple of lines (which I still didn't like duplicating). 
It could be possible to use functions that have the same name in TensorFlow and Numpy, and call the one we need depending on the required backend (but it's not the best for maintainability).

Reply:
No problem. It's something we can target for future enhancements, but not a deal-breaker in terms of this PR. can we add a line near the top of the docstring "Note: currently this function is only supported for the 'fock' backend"?